app/routes/_layout+/kits.$kitId.manage-assets.tsx

```diff
@@ -0,0 +1,510 @@
+        where: { id: kitId, organizationId },
+        include: {
+          assets: { select: { id: true, title: true, kit: true } },
+          custody: {
+            select: { custodian: { select: { id: true, name: true } } },
+          },
+        },
+      })
+      .catch((cause) => {
+        throw new ShelfError({
+          cause,
+          message: "Kit not found",
+          additionalData: { kitId, userId, organizationId },
+          status: 404,
+          label: "Kit",
+        });
+      });
+
+    const removedAssets = kit.assets.filter(
+      (asset) => !assetIds.includes(asset.id)
+    );
+
+    const newlyAddedAssets = await db.asset
+      .findMany({
+        where: { id: { in: assetIds } },
+        select: { id: true, title: true, kit: true, custody: true },
+      })
+      .catch((cause) => {
+        throw new ShelfError({
+          cause,
+          message:
+            "Something went wrong while fetching the assets. Please try again or contact support.",
+          additionalData: { assetIds, userId, kitId },
+          label: "Assets",
+        });
+      });
+
+    /** An asset already in custody cannot be added to a kit */
+    const isSomeAssetInCustody = newlyAddedAssets.some(
+      (asset) => asset.custody && asset.kit?.id !== kit.id
+    );
+    if (isSomeAssetInCustody) {
+      throw new ShelfError({
+        cause: null,
+        message: "Cannot add unavailable asset in a kit.",
+        additionalData: { userId, kitId },
+        label: "Kit",
+      });
+    }
+
+    await db.kit.update({
+      where: { id: kit.id, organizationId },
+      data: {
+        assets: {
+          /**
+           * set: [] will make sure that if any previously selected asset is removed,
+           * then it is also disconnected from the kit
+           */
+          set: [],
+          /**
+           * Then this will update the assets to be whatever user has selected now
+           */
+          connect: newlyAddedAssets.map(({ id }) => ({ id })),
+        },
+      },
+    });
+
+    await createBulkKitChangeNotes({
+      kit,
+      newlyAddedAssets,
+      removedAssets,
+      userId,
+    });
+
+    /**
+     * If a kit is in custody then the assets added to kit will also inherit the status
+     */
+    const assetsToInheritStatus = newlyAddedAssets.filter(
+      (asset) => !asset.custody
+    );
+    if (
+      kit.custody &&
+      kit.custody.custodian.id &&
+      assetsToInheritStatus.length > 0
+    ) {
+      await Promise.all([
+        ...assetsToInheritStatus.map((asset) =>
+          db.asset.update({
+            where: { id: asset.id },
+            data: {
+              status: AssetStatus.IN_CUSTODY,
+              custody: {
+                create: {
+                  custodian: { connect: { id: kit.custody?.custodian.id } },
+                },
+              },
+            },
+          })
+        ),
+        ...assetsToInheritStatus.map((asset) =>
+          createNote({
```

Comment:
Same as in the other file. Lets make a function that uses `createMany` rather than looping and creating 1 by one.